New bulk node light update

juhdanad opened the pull request:
**Based on #4682, please do not merge this before!**
Edit: #4682 is merged.

This is an early version for testing, the code style is not final yet.
Lua changes are not documented yet.
Edit: now the PR is done.

The new lighting algorithm should fix #3421 (mostly, sometimes a well-lit area is overridden, probably by the mapgen). With this v7 mapgen should be playable with the moretrees mod.

This should also fix #4853 and fix #4852.

Changes:
- If the block above the changed area is not loaded, it gets loaded to avoid lighting bugs.
- Light is updated as soon as `write_to_map` is called on a voxel manipulator
- Therefore `update_map` does nothing
- Schematics, trees and voxel manipulators use the new lighting algorithm

Fixer-007 commented:
Will try it...

paramat commented:
This work is very much appreciated, these lighting bugs are one of the biggest problems in Minetest.

juhdanad commented:
Done! The WIP label can be removed.

HybridDog commented:
If many mods add map generation, all of them use the vmanip object and call calc_lighting,
then each time calc_lighting is executed the previous calculation is removed (it was in vain) and light is calculated again, am l wrong?
So instead calc_light could only set a flag indicating that the mapchunk needs light calculation after executing the mods' on_generateds.
Or does this PR remove the calc_light function from vmanip?

juhdanad commented:
@HybridDog yes, it turned out there are now problems with map generating (mods that use singlenode mapgen with nolight flag)
I'll try to fix that.

paramat commented:
> If many mods add map generation, all of them use the vmanip object and call calc_lighting,
then each time calc_lighting is executed the previous calculation is removed (it was in vain) and light is calculated again, am l wrong?

In this case it would be the mods responsibility to delay the light calculation until after the last 'on generated' operation. Each mod can have a bool to disable lighting calculations.

juhdanad commented:
Now I pushed a patch.
The behavior of lighting only changes for not mapgen voxel manipulators.
I have tested with the following mods:

- moretrees
- nuke
- stability
- pathv7
- tnt

I hope this is now free of bugs.

juhdanad commented:
I have also made some speed tests:
Average light calculation time when moretrees places a tree: 125 ms->26 ms
TNT exploding in sunlight (old): 12 ms
TNT exploding in shadow (old): 2,4 ms
TNT exploding (new): 2,1 ms

Fixer-007 commented:
It is still WIP or I can try it?

juhdanad commented:
@Fixer-007 the first commit in itself is #4682 and the second commit is the new stuff. If #4682 gets merged, the first commit will disappear from there.

Fixer-007 commented:
Checked for now:
* fixes incorrect light after tnt in water explosion
* fixes shadows induced by tnt blown over to far horizon
* fixes that L system trees shadows, moretrees are working nice!

Have not checked yet:
* general stuff
* sofars lighting and sfans nuke mod
* ghost light after lava casts fix should work here, since it has that PR included, but I have not checked

High hopes for this.

Fixer-007 commented:
Needs rebase.

juhdanad commented:
Rebased! Please review #4682 first, as merging it will remove two commits from there.

paramat commented:
Unfortunately hmmmm isn't around recently, this needs very careful consideration, because it affects voxel manipulators and 'upate map' i fell hmmmm should look at this.

paramat commented:
Bump! Unexciting but this is top priority for review as it fixes MTs most serious current bug.

Fixer-007 commented:
This is very needed PR, really important, fixes lots of stuff.

MT-Modder commented:
I'm surprised this hasn't been merged...

paramat commented:
Difficult to review, lighting is complex to understand, but i will try to do my best.

MT-Modder commented:
Thanks @paramat

I wish I knew anything about C++ or lighting calculations.

paramat commented:
I have some questions and ideas for this i will discuss later or tomorrow.

HybridDog commented:
ok.

paramat commented:
> Light is updated as soon as write_to_map is called on a voxel manipulator
Therefore update_map does nothing

Now that lighting calculation is moved from 'update_map' into 'write_to_map', how does this affect a mapgen-object voxelmanip that uses 'write_to_map'?
A mapgen-object voxelmanip can assume complete darkness (either by using the 'nolight' flag or doing 'set lighting = 0') and can therefore use 'calc_lighting' which is a faster lighting calculation.
With this PR it seems that a mapgen-object voxelmanip will be doing 'calc_lighting' and then 'write_to_map' which contains a second and unnecessary lighting calculation.
Do we now then need 2 versions of 'write_to_map' and only use the new one in a non-mapgen-object voaxelmanip?

juhdanad commented:
@paramat mapgen voxel manipulators are not affected. In your situation write_to_map does not calculate lighting (see [there](https://github.com/minetest/minetest/pull/4967/files#diff-886a66bda4427538b1f65495b7540f92R115)). This pull is about non-mapgen voxel manipulators only.

paramat commented:
I see, thanks.

juhdanad commented:
I don't recommend reviewing this before #4682 is merged, as I will rebase this pull and the amount of changes will reduce.

paramat commented:
#4682 merged.

juhdanad commented:
Rebased! The amount if changes has decreased.

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+	MapBlock *source_block = map->emergeBlock(source_block_pos, false);
+	// Trust only generated blocks.
+	if (source_block == NULL || source_block->isDummy()
+		|| !source_block->isGenerated()) {
```
One more indent

SmallJoker commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+	// Distance in array between two nodes on top of each other.
+	s16 step=vm->m_area.getExtent().X;
+	// Cache the ignore node.
+	MapNode ignore=MapNode(CONTENT_IGNORE);
```
Spaces (above too)

paramat replied:
Spaces around '='

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+			for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
+				light[x][z] = sunlight;
+			}
+		}
```
```
+		for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+		for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
+			light[x][z] = sunlight;
+		}
```
?

SmallJoker commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+					n = &vm->m_data[i];
+				}
+				// Ignore IGNORE nodes, these are not generated yet.
+				if(n->getContent()==CONTENT_IGNORE)
```
`if (n->getContent() == CONTENT_IGNORE)`

paramat replied:
Spaces around '=='

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+					light[x][z] = false;
+				}
+			}
+		}
```
```
+		for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+		for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
+			// Get the bottom block.
+			MapNode above = source_block->getNodeNoCheck(x, 0, z,
+				&is_valid_position);
+			if (above.getLight(LIGHTBANK_DAY, ndef) == LIGHT_SUN) {
+				light[x][z] = true;
+			} else {
+				light[x][z] = false;
+			}
+		}
```
?

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+
+	// For each map block:
+	for (s16 x = minblock.X; x <= maxblock.X; x++) {
+		for (s16 z = minblock.Z; z <= maxblock.Z; z++) {
```
```
+	for (s16 x = minblock.X; x <= maxblock.X; x++)
+	for (s16 z = minblock.Z; z <= maxblock.Z; z++) {
```
?

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+	bool light[MAP_BLOCKSIZE][MAP_BLOCKSIZE])
+{
+	// Distance in array between two nodes on top of each other.
+	s16 step=vm->m_area.getExtent().X;
```
Spaces around '='

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+	// For each column of nodes:
+	for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+		for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
+			// Position of the column on the map.
```
```
+	for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+	for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
?

paramat replied:
```
+	for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+	for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
?

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+	MapNode ignore=MapNode(CONTENT_IGNORE);
+	// For each column of nodes:
+	for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+		for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
```
+	for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+	for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
?

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+			// True if the current node has sunlight.
+			bool lig = light[x][z];
+			// For each node, downwards:
+			for (s32 i = maxindex; i >= minindex; i-=step) {
```
Spaces around '-='

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+			sunlight = !node_block->getIsUnderground();
+		}
+		for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+			for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
```
+		for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+		for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
?

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+		bool is_valid_position;
+		// For each column:
+		for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+			for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
```
+		for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+		for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
?

HybridDog replied:
swap x and z

juhdanad replied:
Why should I swap them?

paramat replied:
Either way is fine. Occasionally it is best to do ZYX or ZX order but only when there is a reason such as incrementing indexes by +1 along an X-row.

HybridDog replied:
The map blocks are stored in zyx order, so putting x into the inner loop improves cache coherency. You'd also need to change the ```light``` array usage.

juhdanad replied:
Since `z` is the second index of the `light` array I think swapping the two indices would make the coherence worse.

HybridDog replied:
>You'd also need to change the light array usage.

Changing the order would also keep consistency with the order in mapblocks.

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+			// Copy sunlight data
+			data.target_block = v3s16(x, minblock.Y - 1, z);
+			for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+				for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
```
+			for (s16 x = 0; x < MAP_BLOCKSIZE; x++)
+			for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```
?

paramat commented:
Line comments, mostly code style. I understand this a little.

SmallJoker commented:
Tested this pull, works fine with a Lua mapgen and Moretrees - so far no light bugs.
![screenshot_20170219_100723_](https://cloud.githubusercontent.com/assets/1497498/23100947/72c680c2-f68b-11e6-81d4-30187aef35fc.png)

juhdanad commented:
I have corrected the code style! (but I did not swap the `x` and `z` for loops, I don't want to accidentally spoil my code)

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -1055,10 +1047,9 @@ bool propagate_block_sunlight(Map *map, INodeDefManager *ndef,
 					relight->push(LIGHT_SUN, current_pos, data->target_block,
 						block, 4);
-				} else {
+				} else
```
This should be reverted to how it was before, we don't mix bracketed and non-bracketed style in one conditional.

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -1073,10 +1064,9 @@ bool propagate_block_sunlight(Map *map, INodeDefManager *ndef,
 					unlight->push(LIGHT_SUN, current_pos, data->target_block,
 						block, 4);
-				} else {
+				} else
```
As above.

paramat commented:
Added 2 line comments, otherwise commits 2 and 3 look fine.

paramat commented:
Yes changing to ZYX order is potentially good, it's certainly consistent with other code.

paramat commented:
Stress-testing with my catacomb mod, which uses the non-mapgen object voxelmanip to rapidly create structures in the world, light bugs are almost completely fixed. Normally this mod creates large dark shadows. There are a few very subtle shadows but these are likely to be due to the extreme nature of the test.

paramat commented:
I updated the catacomb mod to be used for testing light updates https://github.com/paramat/catacomb

paramat commented:
Also getting bright areas underwater:

![screenshot_20170224_060003](https://cloud.githubusercontent.com/assets/3686677/23292304/bf172c28-fa56-11e6-8480-839d5d728f4f.png)

So unfortunately i think this PR needs work.

paramat commented:
Testing with my projects mod https://github.com/paramat/projects : 

![screenshot_20170224_062310](https://cloud.githubusercontent.com/assets/3686677/23292829/459c11de-fa5a-11e6-8c2b-5d9f1cec1607.png)

^ Seems too bright under this structure

![screenshot_20170224_062336](https://cloud.githubusercontent.com/assets/3686677/23292838/50f0ba44-fa5a-11e6-9495-42845cba07b7.png)

^ Subtle light bugs (nothing above)

paramat commented:
Compare to master:

![screenshot_20170224_063055](https://cloud.githubusercontent.com/assets/3686677/23293014/6cf7d078-fa5b-11e6-9645-43438ce27aeb.png)

![screenshot_20170224_063325](https://cloud.githubusercontent.com/assets/3686677/23293016/70d47aca-fa5b-11e6-994a-01a54a00ca72.png)

This PR seems to not create shadows as dark as they need to be, and shadows are messy.

paramat commented:
I'm a little unconfortable with how this is a rewrite instead of an evolution of existing code.
Rewrites are often not a good idea. Code that has slowly evolved over years (as this has) contains a huge amount of testng and tuning history which is extremely valuable, it's probably better to continue this slow organic evolution.

The change that is certainly needed, and that works in testing, is the loading of blocks above if not loaded. Please can we try a different approach of gradual evolution of lighting code in small steps, and start with adding the loading of blocks above?

paramat commented:
I've learnt a lot from this PR:
It's important to test for yourself, even if others have tested, as people can miss things.
Testing should come before code review, because if the code doesn't work it will need to be changed and the code review is then partly wasted effort.

Zeno- commented:
> Code that has slowly evolved over years (as this has) contains a huge amount of testng and tuning history

And lots of hacks on top of hacks ;)

paramat commented:
I certainly feel it would be a mistake for us to completely rewrite this code. It basically works very well but has a shadow problem. Instead let's discover how the bug happens and make the minimum changes to fix that, this seems to be the loading of mapblocks above the processed volume, please could you make this change to the existing code?

paramat commented:
To test with the projects mod, in creative mode place a "projects:vertenta" spawner node on the ground and wait nearby.

Apart from the loading of blocks above and updating light when 'writing to map', how much is this code built on current code? If it's very different what are the reasons for that?

juhdanad commented:
@paramat yes, I also see the bugs now that I have these tools, thanks for your mods.

I think I also found the reason of them: the 'projects' bug appears because the mod only sets the content of the blocks and does not modify the light information in them - and my code did not check if the node can have light at all.
The 'catacombs' bug is due to the incorrect replacement of `x` and `z` indices in the second commit.
I have corrected these, now I'm trying to find any more bugs...

The original bulk node lighting code is different from the one that I have written: the current uses sets to store currently processed nodes, I use faster vectors.
Also the old code uses `diminishLight()` and `undiminishLight()` which limited the light sources' light incorrectly.
The old code handles water differently than other nodes (see [there](https://github.com/minetest/minetest/blob/master/src/mapblock.cpp#L262)), and is therefore also inconsistent with the mapgen lighting system (this is the reason of the L-system tree lighting bug).
The old code unlights all nodes on the border of the mapblocks, I only do that if the node will not have direct sunlight after the update.
And finally this pull uses my previous lighting PRs to do the light updates efficiently, these are different procedures than the currently used ones. So even if I modified the original lighting code, I would have to rewrite most of it.

paramat commented:
Ok thanks, sorry to be careful about this, i can now see there's good reason to change the code, so no problem as long as it works.

> the mod only sets the content of the blocks and does not modify the light information in them

This is how the lua voxelmanip and schematics are used so has to be accounted for. I can't think of any LVM or schematic usage that sets light values.

juhdanad commented:
There is another rare bug:

1. The mapgen thread reads a new chunk and clones its surroundings
2. The mod (or a player) places a light source near the ungenerated area, and light spreads to a cloned block
3. Map generation is done, and the cloned blocks are copied back, replacing the correct light

This bug is avoidable by checking lighting at the end of the generation, but it might cost some time (but far less than a full light update, as light is unchanged in most cases, and only borders are checked)

@paramat does the map generator always blit back a cuboid-shaped area or are there exceptions?

paramat commented:
As far as i know core mapgen uses a core voxelmanip volume that is the mapchunk plus a mapblock-deep shell of surrounding mapblocks. So by default the voxelmanip volume is always (80 + 16 + 16) ^3 = 112^3 nodes.
Any Lua mapgen using the mapgen-object voxelmanip uses that same volume, the mapgen-object voxelmanip is just a copy of the core voxelmanip in it's state after core mapgen has finished.
Lua mapgens should only place light sources in the mapchunk, and because light never spreads more than 16 nodes it can't reach to beyond the 112^3 node volume.

juhdanad commented:
I have to admit that I made some mistakes. I have corrected everything I found. I hope that you will be satisfied now.

But please keep in mind that the underwater light patches are correct, since stairs let light trough.
Because of the additional lighting bug check map chunk generation is ~0.12 seconds longer (tested on a debug build, should be even less on a release build). I hope that it is acceptable.

paramat commented:
> underwater light patches are correct, since stairs let light trough.

I see, i should change my mod code to place full nodes under the stairs.

> map chunk generation is ~0.12 seconds longer

This is concerning, for core mapgen or a non-mapgen-object lua voxelmanip acting on a mapchunk-sized area?
What is this additional lighting bug check for? Is the bug common?

HybridDog commented:
Please do not combine write_to_map and update_map, in my Worldedit fork l write all nodes at once and gradually update the map where needed: 
https://github.com/HybridDog/WorldEdit-pack/blob/hversion/worldedit/common.lua#L108
This way the performance is increased tremendously.

paramat commented:
That's one usage by one person though, the combination of those 2 functions greatly increaases performance for many parts of Minetest.

paramat commented:
I'll test this without the changes to map.cpp in commit 5.

juhdanad commented:
@paramat the changes in map.cpp are fixing a rare bug: when the surroundings of a generating map chunk are modified during the generation. To reproduce this bug, generate chambers of light sources and fly towards an ungenerated area. If you are lucky, a chamber is generated in the right place and time, and you will see a sudden cut in the spreading light.

@HybridDog why would your mod become slow? Are your voxel manipulators huge? Or do you update the same area often? I just wanted to combine the two methods because they are faster that way, but I can keep the old way too.

paramat commented:
Ok so the code in map.cpp is something run when core mapgen generates?
It's worth timing this in a release build, 0.12s added to core mapgen mapchunk generation time is totally unacceptable for a rarely-triggered bug.
Usually usages of a non-mapgen-object-vm are not near generating mapchunks, i think my catacomb mod has an unreasonably fast generation sometimes, something to do with how the ABMs behave.

paramat commented:
In case you want to test i have updated my 'catacomb' mod to place full nodes under the stairs to block the light. I will retest your PR soon, with 'projects' mod too.

HybridDog commented:
> [Why] would your mod become slow?

Worldedit allows setting lots of nodes at once. If someone sets a 100x100x100 cube made of mese it takes ~0.05 s with my Worldedit and ~5.78 s with the original one, the current update_map takes lots of time. My version uses my function_delayer mod to distribute necessary map updates over time without causing noticeable server lag. By updating each chunk separately, l can omit unloaded chunks, which may be the majority, by using minetest.get_node_or_nil.

> Are your voxel manipulators huge? 

If players set lots of nodes, they're huge. If they set a few nodes, the overhead is negligible.

>[Do] you update the same area often?

Yes, often players use the command frequently to change the nodes e.g. just because they noticed that cobble, then stone brick looked more fitting after all. If the map updates are delayed, no redundant ones are executed.

Is updating the map after your change still about 100 times faster than writing to map?

juhdanad commented:
@HybridDog the current state is:

- write_to_map automatically corrects light (but light correction is faster than before)
- update_map does nothing

I will see how can I solve this problem, as simply disabling light updates is not enough: all of my lighting methods assume that the lighting on the map is correct (this gives some extra speed) and by placing incorrect light on the map it is not guaranteed that bugs will disappear when you update the light again.

HybridDog commented:
thanks

paramat commented:
I'm doubt it's worth altering your PR for this single very specialist usage.

juhdanad commented:
@paramat I have tested the post-mapgen light update on a release build, it takes 0,024 seconds.
However, I removed that as you asked. I hope that this pull is finished now.

@HybridDog I will open a new PR for you, I hope that you understand that this PR is complex enough already.

juhdanad commented:
@HybridDog there are the corrections for you, I hope they will be sufficient: https://github.com/juhdanad/minetest/commit/23fdd16fe20b82d1c1289d03af01c89764ac4be6
With this, you can disable light updating during `write_to_map()` and call `minetest.fix_light()` later (new function, repairs the light in one map block)
I will open a PR when this is mered.

HybridDog commented:
thanks, does write_to_map still work about 100 time faster without light fixing?

juhdanad commented:
@HybridDog I do not know how fast `write_to_map` without light update is. But light update is about 3-4 times faster than before.

paramat commented:
Thanks, will test.
24ms is quite significant, i think too much for a rare bug. Anyway that can be considered in a separate PR.

paramat commented:
Retested with 'catacomb' and 'projects' mods, all seems good now :+1:

sofar commented:
:+1: tested with large voxelmanips.

Any remaining issues can be ironed out later - I think this shouldn't wait much longer.

MarkuBu commented:
Great. Than I finally can add l-system trees?

paramat commented:
ab371cc93491baf0973ecc94b96c3a1fdb4abfd5
Hero.
